src/client/Pipeline.cpp

```diff
@@ -241,18 +271,41 @@ void PipelineImpl::buildForAppendOrRecovery(bool recovery) {
              */
             if (stage != PIPELINE_SETUP_CREATE && stage != PIPELINE_CLOSE
                     && static_cast<int>(nodes.size()) < replication && canAddDatanode) {
-                if (!addDatanodeToPipeline(excludedNodes)) {
+                // Single data node case
+                bool added = false;
+                if (nodes.empty() && useRemoved) {
+                    if (storageID.length()) {
+                        LOG(INFO, "Pipeline: Adding back only datanode %s", removed.formatAddress().c_str());
+                        nodes.push_back(removed);
```

Comment:
Ok, I see what you are saying now about the end of the function. When I tested this in single data mode it did work like this, so I don't think it is a problem. Also, the alternate addDatanodeToPipeline also modifies the block in a different way so I think this would be equivalent.